Add doc for SCSt function bindings in SCDF (#236)
spring-io/dataflow.spring.io: content/documentation/pages/6-recipes/8-functional-apps/1-scst-function-bindings.md

---
path: 'recipes/functional-apps/scst-function-bindings/'
title: 'Functional Applications'
description: 'Configuring the Spring Cloud Stream Functional applications'
---

> # Building Streaming Pipeline using Functional applications

Reviewer: ```suggestion
# Building Streaming Data Pipeline using Functional applications
```

> To learn how to build simple Function based applications using Spring Cloud Stream and have them embedded into Spring Cloud Data Flow Streaming pipeline.

Reviewer: ```suggestion
To learn how to build simple Function-based applications using Spring Cloud Stream and have them embedded into the Spring Cloud Data Flow Streaming data pipeline.
```

## Overview

> Let's create a simple source `time-source` that emits Date Formats at a configured interval to the messaging middleware and the sink `log-sink` consumes the published messages from the middleware.

Reviewer: ```suggestion
Let's create a `time-source` application that produces the current date/timestamp at a configured interval to the messaging middleware, and the sink `log-sink` consumes the published messages from the middleware.
```

The following sample applications are available in Spring Cloud Data Flow samples [repo](https://github.com/spring-cloud/spring-cloud-dataflow-samples/tree/master/spring-cloud-stream-function-bindings).

With Spring Cloud Stream 3.x as the dependency, you can write the `Source` application using `java.util.function.Supplier` as follows:

```
package com.example.timesource;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.function.Supplier;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@SpringBootApplication
public class TimeSourceApplication {

	public static void main(String[] args) {
		SpringApplication.run(TimeSourceApplication.class, args);
	}

	@Configuration
	public class TimeSupplierConfiguration {
		public TimeSupplierConfiguration() {
		}

		@Bean
		public Supplier<String> timeSupplier() {
			SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");
			return () -> {
				return sdf.format(new Date());
			};
		}
	}

> }

Author: Yes, that will make it better @sobychacko

import org.springframework.context.annotation.Configuration;
import org.springframework.integration.dsl.IntegrationFlow;
import org.springframework.integration.dsl.IntegrationFlows;
import org.springframework.integration.handler.LoggingHandler;
import org.springframework.messaging.Message;

@Configuration
public class LogConsumerConfiguration {
	public LogConsumerConfiguration() {
	}

	@Bean
	IntegrationFlow logConsumerFlow() {
		return IntegrationFlows.from(MessageConsumer.class, (gateway) -> gateway.beanName("logConsumer"))
				.handle((payload, headers) -> {
					if (payload instanceof byte[]) {
						return new String((byte[]) payload);
					}
					return payload;
				})
				.log(LoggingHandler.Level.INFO, "log-consumer", "payload")
				.get();
	}

	private interface MessageConsumer extends Consumer<Message<?>> {}
}

```

and,

```
package com.example.logsink;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class LogSinkApplication {

	public static void main(String[] args) {
		SpringApplication.run(LogSinkApplication.class, args);
	}

> }

Author: That's a good idea. I will get this fixed as well.

```

After building both the `time-source` and `log-sink` applications, you can register these applications into Spring Cloud Data Flow.

Let's say you create a stream using these applications as follows:

```
ticktock=time-source | log-sink
```

You need to make sure that the functional bindings of these applications are mapped to the appropriate `output` and `input` names that Spring Cloud Data Flow understands.

> What this means is that, you need to configure the following properties when `deploying` the stream:

Reviewer: ```suggestion
What this means is that you need to configure the following properties when deploying the stream:
```

```
app.time-source.spring.cloud.stream.function.bindings.timeSupplier-out-0=output
app.log-sink.spring.cloud.stream.function.bindings.logConsumer-in-0=input
```

The timeSupplier function's output and `logConsumer` function's input need to be mapped to the outbound/inbound names that Spring Cloud Data Flow understands: `output` and `input`.

> Along with this, you also need to provide a way trigger the `Supplier` function. In this case,

Reviewer: Maybe move this to the DSL itself via `--fixed-delay=5`? One less step to discuss in the demo walkthrough.

Author: Since I didn't add the properties white-listing, I stuck to this approach. Otherwise, we would need to add the white-listing into the samples. What do you think?

Reviewer: OK, let's go without it.
